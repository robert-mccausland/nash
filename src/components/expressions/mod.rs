use crate::{
    executor::{ExecutorContext, ExecutorStack, Value},
    lexer::{Token, TokenValue},
    utils::iterators::Backtrackable,
};

mod block;
mod branch;
mod collections;
mod execute;
mod loops;
mod variable;

use super::{
    block::Block,
    errors::ParserError,
    literals::{BooleanLiteral, CommandLiteral, IntegerLiteral, StringLiteral},
    operator::Operator,
    EvaluationResult, Tokens,
};

use block::BlockExpression;
use branch::BranchExpression;
use collections::{ArrayExpression, TupleExpression};
use execute::ExecuteExpression;
use loops::{ForLoopExpression, WhileLoopExpression};
use serde::Serialize;
use variable::VariableExpression;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Expression {
    pub operations: Vec<(Operator, BaseExpression)>,
    pub first: BaseExpression,
}

impl Expression {
    pub fn new(first: BaseExpression, operations: Vec<(Operator, BaseExpression)>) -> Self {
        Self {
            first: first.into(),
            operations,
        }
    }

    pub(super) fn parse<'a, I: Iterator<Item = &'a Token<'a>>>(
        tokens: &mut Backtrackable<I>,
    ) -> Result<Expression, ParserError> {
        let expression = BaseExpression::parse(tokens)?;
        let mut operations = Vec::new();
        while let Some(operator) = Operator::try_parse(tokens) {
            operations.push((operator, BaseExpression::parse(tokens)?));
        }

        return Ok(Expression::new(expression, operations));
    }

    pub fn evaluate(
        &self,
        stack: &mut ExecutorStack,
        context: &mut ExecutorContext,
    ) -> EvaluationResult<Value> {
        let mut result = self.first.evaluate(stack, context)?;

        for (operator, expression) in &self.operations {
            let right = expression.evaluate(stack, context)?;
            result = operator.execute(result, right)?;
        }
        return Ok(result.into());
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct BaseExpression {
    content: Box<ExpressionContent>,
    accessor: Option<u32>,
}

impl BaseExpression {
    fn try_parse_accessor<'a, I: Iterator<Item = &'a Token<'a>>>(
        tokens: &mut Backtrackable<I>,
    ) -> Option<u32> {
        tokens.backtrack_if_none(|tokens| {
            let Some(TokenValue::Dot()) = tokens.next_value() else {
                return None;
            };

            let Some(TokenValue::IntegerLiteral(integer)) = tokens.next_value() else {
                return None;
            };

            return u32::from_str_radix(integer, 10).into_iter().next();
        })
    }
}

impl BaseExpression {
    fn parse<'a, I: Iterator<Item = &'a Token<'a>>>(
        tokens: &mut Backtrackable<I>,
    ) -> Result<Self, ParserError> {
        let content = ExpressionContent::parse(tokens)?;
        let accessor = Self::try_parse_accessor(tokens);
        return Ok(BaseExpression {
            accessor,
            content: content.into(),
        });
    }

    fn evaluate(
        &self,
        stack: &mut ExecutorStack,
        context: &mut ExecutorContext,
    ) -> EvaluationResult<Value> {
        let value = self.content.evaluate(stack, context)?;

        if let Some(accessor) = self.accessor {
            let Value::Tuple(mut values) = value else {
                return Err("Cannot use get expression on non-tuple value".into());
            };

            let len = values.len();
            let result = values.get_mut(accessor as usize).ok_or(format!(
                "Cannot get element at index {:} because tuple only has {:} elements",
                accessor, len
            ))?;

            Ok(core::mem::take(result).into())
        } else {
            Ok(value.into())
        }
    }
}

macro_rules! expression_content {
    ([$($expression_type:ident,)*]) => {
        use crate::components::Evaluatable;
        use crate::components::Parsable;

        #[derive(Debug, Clone, PartialEq, Eq, Serialize)]
        enum ExpressionContent {
            $(
                $expression_type($expression_type),
            )*
        }

        impl ExpressionContent {
            fn parse<'a, I: Iterator<Item = &'a Token<'a>>>(
                tokens: &mut Backtrackable<I>,
            ) -> Result<Self, ParserError> {
                $(
                    if let Some(value) = $expression_type::try_parse(tokens)? {
                        return Ok(ExpressionContent::$expression_type(value));
                    }
                )*

                return Err("Unable to parse expression".into());
            }

            fn evaluate(
                &self,
                stack: &mut ExecutorStack,
                context: &mut ExecutorContext,
            ) -> EvaluationResult<Value> {
                match self {
                    $(
                        Self::$expression_type(value) => value.evaluate(stack, context),
                    )*
                }
            }
        }
    };
}

expression_content!([
    StringLiteral,
    BooleanLiteral,
    IntegerLiteral,
    CommandLiteral,
    ArrayExpression,
    TupleExpression,
    VariableExpression,
    ExecuteExpression,
    WhileLoopExpression,
    ForLoopExpression,
    BranchExpression,
    BlockExpression,
]);
