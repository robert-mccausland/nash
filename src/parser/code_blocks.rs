use crate::{
    constants::FUNC,
    lexer::{Token, TokenValue},
};

use super::{
    functions::{parse_function, Function},
    statements::{parse_statement, Statement},
    Backtrackable, ParserError,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CodeBlock {
    pub statements: Vec<Statement>,
    pub functions: Vec<Function>,
}

pub(super) fn parse_code_block<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
    root: bool,
) -> Result<CodeBlock, ParserError> {
    let mut statements = Vec::new();
    let mut functions = Vec::new();

    if !root {
        let Some(TokenValue::LeftCurly()) = tokens.next().map(|x| x.value) else {
            return Err("code block must start with {".into());
        };
    }

    loop {
        let token = tokens.next().map(|x| x.value);
        if !root {
            if let Some(TokenValue::RightCurly()) = token {
                break;
            };
        }
        if token.is_none() {
            if root {
                break;
            } else {
                return Err("unexpected end of file".into());
            }
        }

        if let Some(TokenValue::Keyword(FUNC)) = token {
            functions.push(parse_function(tokens)?);
        } else {
            tokens.backtrack();
            statements.push(parse_statement(tokens)?);
        }
    }

    return Ok(CodeBlock {
        statements,
        functions,
    });
}
