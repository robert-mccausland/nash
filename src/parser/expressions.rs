use crate::{
    keywords::VAR,
    lexer::{Token, TokenValue},
};

use super::{
    code_blocks::{parse_code_block, CodeBlock},
    literals::{parse_integer, parse_string_literal, StringLiteral},
    operators::{parse_operator, Operator},
    Backtrackable, Identifier, ParserError,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expression {
    StringLiteral(StringLiteral),
    BooleanLiteral(bool),
    IntegerLiteral(i32),
    Tuple(Vec<Expression>),
    Variable(Identifier),
    Command(Vec<StringLiteral>),
    Operation(Box<Expression>, Operator, Box<Expression>),
    If(Vec<(Expression, CodeBlock)>, Option<CodeBlock>),
    FunctionCall(Identifier, Vec<Expression>),
    Execute(Box<Expression>, Option<CaptureExitCode>),
}

#[derive(Debug, Clone, PartialEq, Eq)]

pub enum CaptureExitCode {
    Assignment(Identifier),
    Declaration(Identifier),
}

pub(super) fn parse_expression<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Result<Expression, ParserError> {
    let mut expression = parse_base_expression(tokens)?;
    while let Some(operator) = parse_operator(tokens) {
        expression = Expression::Operation(
            Box::new(expression),
            operator,
            Box::new(parse_base_expression(tokens)?),
        );
    }

    return Ok(expression);
}

fn parse_base_expression<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Result<Expression, ParserError> {
    let token = tokens.next().map(|x| x.value);
    if let Some(TokenValue::IntegerLiteral(literal)) = token {
        return Ok(Expression::IntegerLiteral(parse_integer(literal)?));
    }

    if let Some(TokenValue::DoubleQuote()) = token {
        return Ok(Expression::StringLiteral(parse_string_literal(tokens)?));
    }

    if let Some(TokenValue::Backtick()) = token {
        let mut command = Vec::new();
        loop {
            let next = tokens.next().map(|x| x.value);
            command.push(if let Some(TokenValue::DoubleQuote()) = next {
                parse_string_literal(tokens)?
            } else if let Some(TokenValue::StringLiteral(value)) = next {
                value.into()
            } else if let Some(TokenValue::Backtick()) = next {
                break;
            } else {
                return Err("Unable to parse command".into());
            });
        }

        return Ok(Expression::Command(command));
    }

    if let Some(TokenValue::Identifier(identifier)) = token {
        let identifier = Identifier {
            value: identifier.to_owned(),
        };

        if matches!(
            tokens.next().map(|x| x.value),
            Some(TokenValue::LeftBracket())
        ) {
            let mut args = Vec::new();
            let mut next = tokens.next().map(|x| x.value);
            while !matches!(next, Some(TokenValue::RightBracket())) {
                tokens.backtrack();
                args.push(parse_expression(tokens)?);
                next = tokens.next().map(|x| x.value);

                if !matches!(
                    next,
                    Some(TokenValue::RightBracket()) | Some(TokenValue::Comma())
                ) {
                    return Err("Expected function argument to be followed by `,` or `)`".into());
                }
            }

            return Ok(Expression::FunctionCall(identifier, args));
        } else {
            tokens.backtrack();
            return Ok(Expression::Variable(identifier));
        }
    }

    if let Some(TokenValue::Keyword(ref keyword)) = token {
        if *keyword == "if" {
            let expression = parse_expression(tokens)?;
            let block = parse_code_block(tokens, false)?;
            if let Some(TokenValue::Keyword("else")) = tokens.next().map(|x| x.value) {
                return Ok(Expression::If(
                    vec![(expression, block)],
                    Some(parse_code_block(tokens, false)?),
                ));
            } else {
                tokens.backtrack();
                return Ok(Expression::If(vec![(expression, block)], None));
            }
        }
        if *keyword == "exec" {
            let expression = parse_expression(tokens)?;
            if let Some(TokenValue::Question()) = tokens.next().map(|x| x.value) {
                let token = tokens.next().map(|x| x.value);
                if let Some(TokenValue::Keyword(VAR)) = token {
                    let Some(TokenValue::Identifier(identifier)) = tokens.next().map(|x| x.value)
                    else {
                        return Err("var must be followed by identifier".into());
                    };
                    return Ok(Expression::Execute(
                        Box::new(expression),
                        Some(CaptureExitCode::Declaration(identifier.into())),
                    ));
                } else if let Some(TokenValue::Identifier(identifier)) = token {
                    return Ok(Expression::Execute(
                        Box::new(expression),
                        Some(CaptureExitCode::Assignment(identifier.into())),
                    ));
                } else {
                    return Err("? must be followed by an var or identifier".into());
                }
            } else {
                tokens.backtrack();
                return Ok(Expression::Execute(Box::new(expression), None));
            }
        }

        if *keyword == "false" {
            return Ok(Expression::BooleanLiteral(false));
        }

        if *keyword == "true" {
            return Ok(Expression::BooleanLiteral(true));
        }
    }

    if let Some(TokenValue::LeftBracket()) = token {
        let mut expressions = Vec::new();
        if let Some(TokenValue::RightBracket()) = tokens.next().map(|x| x.value) {
        } else {
            tokens.backtrack();
            loop {
                expressions.push(parse_expression(tokens)?);

                let next = tokens.next().map(|x| x.value);
                let Some(TokenValue::Comma()) = next else {
                    if let Some(TokenValue::RightBracket()) = next {
                        break;
                    }
                    return Err("Expected , or ) after tuple value".into());
                };
            }
        }

        return Ok(Expression::Tuple(expressions));
    }

    return Err("Unable to parse expression".into());
}
