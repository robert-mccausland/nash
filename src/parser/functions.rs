use crate::lexer::{Token, TokenValue};

use super::{
    code_blocks::{parse_code_block, CodeBlock},
    Backtrackable, Identifier, ParserError,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Function {
    pub name: Identifier,
    pub arguments: Vec<Identifier>,
    pub code: CodeBlock,
}

pub(super) fn parse_function<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Result<Function, ParserError> {
    let Some(TokenValue::Identifier(identifier)) = tokens.next().map(|x| x.value) else {
        return Err("func must be followed by an identifier".into());
    };
    let name = Identifier {
        value: identifier.to_owned(),
    };

    let Some(TokenValue::LeftBracket()) = tokens.next().map(|x| x.value) else {
        return Err("arguments must be followed by )".into());
    };

    // TODO implement function arguments

    let Some(TokenValue::RightBracket()) = tokens.next().map(|x| x.value) else {
        return Err("arguments must be followed by )".into());
    };
    let code = parse_code_block(tokens, false)?;

    return Ok(Function {
        arguments: Vec::new(),
        name,
        code,
    });
}
