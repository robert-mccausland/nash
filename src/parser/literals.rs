use unicode_segmentation::UnicodeSegmentation;

use crate::lexer::{Token, TokenValue};

use super::{Backtrackable, Identifier, ParserError};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StringLiteral {
    pub parts: Vec<(String, Identifier)>,
    pub end: String,
}

impl StringLiteral {
    pub fn new(parts: Vec<(String, Identifier)>, end: String) -> Self {
        Self { parts, end }
    }
}

impl From<String> for StringLiteral {
    fn from(value: String) -> Self {
        StringLiteral::new(vec![], value)
    }
}

impl From<&str> for StringLiteral {
    fn from(value: &str) -> Self {
        value.to_owned().into()
    }
}

pub(super) fn parse_string_literal<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Result<StringLiteral, ParserError> {
    let mut variables = Vec::new();
    let mut end = String::new();

    loop {
        let next = tokens.next().map(|x| x.value);
        if let Some(TokenValue::DoubleQuote()) = next {
            break;
        } else if let Some(TokenValue::StringLiteral(value)) = next {
            end = parse_string(value)?;
        } else if let Some(TokenValue::Dollar()) = next {
            let Some(TokenValue::LeftCurly()) = tokens.next().map(|x| x.value) else {
                return Err("Expected { after $ in string".into());
            };
            let Some(TokenValue::Identifier(value)) = tokens.next().map(|x| x.value) else {
                return Err("Expected identifier after ${ in string".into());
            };
            let Some(TokenValue::RightCurly()) = tokens.next().map(|x| x.value) else {
                return Err("Expected } after template variable".into());
            };
            variables.push((end, value.into()));
            end = String::new();
        } else {
            return Err("Unable to parse string literal".into());
        }
    }

    return Ok(StringLiteral::new(variables, end));
}

pub(super) fn parse_integer(value: &str) -> Result<i32, ParserError> {
    i32::from_str_radix(value, 10)
        .map_err(|_| format!("Unable to parse {value} as a number").into())
}

fn parse_string(value: &str) -> Result<String, ParserError> {
    let mut result = String::new();
    let mut escape = false;
    for grapheme in value.graphemes(true) {
        if grapheme == "\\" && !escape {
            escape = true;
        } else {
            result += grapheme;
            escape = false;
        }
    }

    if escape {
        return Err(format!("Unterminated escape sequence in \"{value}\"").into());
    }

    return Ok(result);
}
