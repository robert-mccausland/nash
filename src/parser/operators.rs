use crate::lexer::{Token, TokenValue};

use super::Backtrackable;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Operator {
    LessThan,
    GreaterThan,
    Addition,
    Equals,
    Pipe,
}

pub(super) fn parse_operator<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Option<Operator> {
    let Some(next) = tokens.next().map(|x| x.value) else {
        return None;
    };

    if let TokenValue::LeftAngle() = next {
        return Some(Operator::LessThan);
    }

    if let TokenValue::RightAngle() = next {
        return Some(Operator::GreaterThan);
    }

    if let TokenValue::Plus() = next {
        return Some(Operator::Addition);
    }

    if let TokenValue::Equals() = next {
        let next = tokens.next().map(|x| x.value);
        if let Some(TokenValue::RightAngle()) = next {
            return Some(Operator::Pipe);
        }

        if let Some(TokenValue::Equals()) = next {
            return Some(Operator::Equals);
        }
        tokens.backtrack();
    }

    tokens.backtrack();
    return None;
}
