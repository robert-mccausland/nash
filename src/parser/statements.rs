use crate::lexer::{Token, TokenValue};

use super::{
    expressions::{parse_expression, Expression},
    Backtrackable, Identifier, ParserError,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Statement {
    Declaration(Assignment, Expression),
    Assignment(Assignment, Expression),
    Expression(Expression),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Assignment {
    Simple(Identifier),
    Tuple(Vec<Identifier>),
}

pub(super) fn parse_statement<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Result<Statement, ParserError> {
    let statement = parse_statement_content(tokens)?;
    let Some(TokenValue::Semicolon()) = tokens.next().map(|x| x.value) else {
        return Err("statement must end with ;".into());
    };
    return Ok(statement);
}

pub(super) fn parse_statement_content<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Result<Statement, ParserError> {
    let next = tokens.next().map(|x| x.value);
    // Any statement starting with var must be a declaration
    if let Some(TokenValue::Keyword("var")) = next {
        let assignment = try_parse_assignment(tokens)
            .ok_or::<ParserError>("var must be followed by an assignment".into())?;
        return Ok(Statement::Declaration(
            assignment,
            parse_expression(tokens)?,
        ));
    }
    tokens.backtrack();

    if let Some(assignment) = try_parse_assignment(tokens) {
        return Ok(Statement::Assignment(assignment, parse_expression(tokens)?));
    }

    // Otherwise it might be a bare expression
    let expression = parse_expression(tokens)?;
    return Ok(Statement::Expression(expression));
}

fn try_parse_assignment<'a, I: Iterator<Item = Token<'a>>>(
    tokens: &mut Backtrackable<I>,
) -> Option<Assignment> {
    let next = tokens.next().map(|x| x.value);

    if let Some(TokenValue::Identifier(identifier)) = next {
        return if let Some(TokenValue::Equals()) = tokens.next().map(|x| x.value) {
            Some(Assignment::Simple(identifier.into()))
        } else {
            tokens.backtrack_n(2);
            None
        };
    };

    if let Some(TokenValue::LeftBracket()) = next {
        let mut identifiers = Vec::new();
        let mut next = tokens.next().map(|x| x.value);
        if let Some(TokenValue::RightBracket()) = next {
        } else {
            let mut n = 0;
            loop {
                let Some(TokenValue::Identifier(identifier)) = next else {
                    tokens.backtrack_n(n);
                    return None;
                };
                identifiers.push(identifier.into());
                next = tokens.next().map(|x| x.value);
                n += 1;
                let Some(TokenValue::Comma()) = next else {
                    if let Some(TokenValue::RightBracket()) = next {
                        break;
                    }
                    tokens.backtrack_n(n);
                    return None;
                };
                next = tokens.next().map(|x| x.value);
                n += 1;
            }
        }
        return if let Some(TokenValue::Equals()) = tokens.next().map(|x| x.value) {
            Some(Assignment::Tuple(identifiers))
        } else {
            tokens.backtrack();
            None
        };
    };

    tokens.backtrack();
    return None;
}
