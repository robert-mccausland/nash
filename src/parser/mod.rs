use std::{error::Error, fmt::Display};

use code_blocks::{parse_code_block, CodeBlock};

use crate::lexer::Token;

pub mod code_blocks;
pub mod expressions;
pub mod functions;
pub mod literals;
pub mod operators;
pub mod statements;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Identifier {
    pub value: String,
}

impl From<&str> for Identifier {
    fn from(value: &str) -> Self {
        Identifier {
            value: value.to_owned(),
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct ParserError {
    message: String,
}

impl ParserError {
    fn new(message: String) -> Self {
        Self { message }
    }
}

impl From<&str> for ParserError {
    fn from(value: &str) -> Self {
        ParserError::new(value.to_owned())
    }
}

impl From<String> for ParserError {
    fn from(value: String) -> Self {
        ParserError::new(value)
    }
}

impl Display for ParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}

impl Error for ParserError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        None
    }

    fn description(&self) -> &str {
        "description() is deprecated; use Display"
    }

    fn cause(&self) -> Option<&dyn Error> {
        self.source()
    }
}

struct Backtrackable<I: Iterator>
where
    I::Item: Clone,
{
    position: usize,
    history: Vec<I::Item>,
    source: I,
}

impl<I: Iterator> Backtrackable<I>
where
    I::Item: Clone,
{
    fn new(source: I) -> Self {
        Self {
            position: 0,
            source,
            history: Vec::new(),
        }
    }

    fn backtrack(&mut self) {
        self.backtrack_n(1);
    }

    fn backtrack_n(&mut self, n: usize) {
        self.position += n;

        if self.position > self.history.len() {
            panic!("Cannot backtrack past the start of the iterator")
        }
    }

    fn peek(&mut self) -> Option<I::Item> {
        let next = self.next();
        self.backtrack();
        return next;
    }
}

impl<I: Iterator> Iterator for Backtrackable<I>
where
    I::Item: Clone,
{
    type Item = I::Item;

    fn next(&mut self) -> Option<Self::Item> {
        if self.position == 0 {
            if let Some(next) = self.source.next() {
                self.history.push(next.clone());
                Some(next)
            } else {
                None
            }
        } else {
            let next = self.history[self.history.len() - self.position].clone();
            self.position -= 1;
            Some(next)
        }
    }
}

pub fn parse<'a>(tokens: impl IntoIterator<Item = Token<'a>>) -> Result<CodeBlock, ParserError> {
    let tokens = &mut Backtrackable::new(tokens.into_iter());
    match parse_code_block(tokens, true) {
        Ok(code) => Ok(code),
        Err(err) => {
            println!("{:?}", tokens.peek());
            Err(err)
        }
    }
}

#[cfg(test)]
mod tests {
    use code_blocks::CodeBlock;
    use expressions::Expression;
    use functions::Function;
    use literals::StringLiteral;
    use operators::Operator;
    use statements::{Assignment, Statement};

    use crate::{
        constants::{EXEC, FUNC, IF, TRUE, VAR},
        lexer::{FilePosition, TokenValue},
    };

    use super::*;

    #[test]
    fn should_parse_valid_tokens() {
        let tokens = vec![
            TokenValue::Keyword(FUNC),
            TokenValue::Identifier("main"),
            TokenValue::LeftBracket(),
            TokenValue::RightBracket(),
            TokenValue::LeftCurly(),
            TokenValue::Keyword(VAR),
            TokenValue::Identifier("template"),
            TokenValue::Equals(),
            TokenValue::DoubleQuote(),
            TokenValue::StringLiteral("cheese"),
            TokenValue::DoubleQuote(),
            TokenValue::Semicolon(),
            TokenValue::Keyword(VAR),
            TokenValue::Identifier("test_identifier"),
            TokenValue::Equals(),
            TokenValue::DoubleQuote(),
            TokenValue::StringLiteral("Blue \\\""),
            TokenValue::Dollar(),
            TokenValue::LeftCurly(),
            TokenValue::Identifier("template"),
            TokenValue::RightCurly(),
            TokenValue::StringLiteral("\\\" and rice!"),
            TokenValue::DoubleQuote(),
            TokenValue::Semicolon(),
            TokenValue::Keyword(IF),
            TokenValue::IntegerLiteral("1"),
            TokenValue::Plus(),
            TokenValue::IntegerLiteral("1"),
            TokenValue::Equals(),
            TokenValue::Equals(),
            TokenValue::IntegerLiteral("2"),
            TokenValue::LeftCurly(),
            TokenValue::Identifier("out"),
            TokenValue::LeftBracket(),
            TokenValue::Identifier("test_identifier"),
            TokenValue::RightBracket(),
            TokenValue::Semicolon(),
            TokenValue::RightCurly(),
            TokenValue::Semicolon(),
            TokenValue::Keyword(EXEC),
            TokenValue::Backtick(),
            TokenValue::StringLiteral("echo"),
            TokenValue::StringLiteral("something"),
            TokenValue::Backtick(),
            TokenValue::Semicolon(),
            TokenValue::RightCurly(),
            TokenValue::Identifier("main"),
            TokenValue::LeftBracket(),
            TokenValue::RightBracket(),
            TokenValue::Semicolon(),
        ];

        let expected_syntax_tree = CodeBlock {
            functions: vec![Function {
                name: "main".into(),
                arguments: vec![],
                code: CodeBlock {
                    functions: vec![],
                    statements: vec![
                        Statement::Declaration(
                            Assignment::Simple("template".into()),
                            Expression::StringLiteral("cheese".into()),
                        ),
                        Statement::Declaration(
                            Assignment::Simple("test_identifier".into()),
                            Expression::StringLiteral(StringLiteral::new(
                                vec![("Blue \"".to_owned(), "template".into())],
                                "\" and rice!".to_owned(),
                            )),
                        ),
                        Statement::Expression(Expression::If(
                            vec![(
                                Expression::Operation(
                                    Box::new(Expression::Operation(
                                        Box::new(Expression::IntegerLiteral(1)),
                                        Operator::Addition,
                                        Box::new(Expression::IntegerLiteral(1)),
                                    )),
                                    Operator::Equals,
                                    Box::new(Expression::IntegerLiteral(2)),
                                ),
                                CodeBlock {
                                    functions: vec![],
                                    statements: vec![Statement::Expression(
                                        Expression::FunctionCall(
                                            "out".into(),
                                            vec![Expression::Variable("test_identifier".into())],
                                        ),
                                    )],
                                },
                            )],
                            None,
                        )),
                        Statement::Expression(Expression::Execute(
                            Box::new(Expression::Command(vec!["echo".into(), "something".into()])),
                            None,
                        )),
                    ],
                },
            }],
            statements: vec![Statement::Expression(Expression::FunctionCall(
                "main".into(),
                vec![],
            ))],
        };

        parser_test(tokens, Ok(expected_syntax_tree));
    }

    #[test]
    fn should_parse_tuple_declaration() {
        let tokens = vec![
            TokenValue::Keyword("var"),
            TokenValue::LeftBracket(),
            TokenValue::Identifier("one"),
            TokenValue::Comma(),
            TokenValue::Identifier("two"),
            TokenValue::RightBracket(),
            TokenValue::Equals(),
            TokenValue::LeftBracket(),
            TokenValue::DoubleQuote(),
            TokenValue::StringLiteral("hi"),
            TokenValue::DoubleQuote(),
            TokenValue::Comma(),
            TokenValue::IntegerLiteral("123"),
            TokenValue::RightBracket(),
            TokenValue::Semicolon(),
        ];

        let expected_tree = CodeBlock {
            functions: vec![],
            statements: vec![Statement::Declaration(
                Assignment::Tuple(vec!["one".into(), "two".into()]),
                Expression::Tuple(vec![
                    Expression::StringLiteral("hi".into()),
                    Expression::IntegerLiteral(123),
                ]),
            )],
        };

        parser_test(tokens, Ok(expected_tree));
    }

    #[test]
    fn should_parse_exec_exit_code_capture() {
        let tokens = vec![
            TokenValue::Keyword(EXEC),
            TokenValue::Backtick(),
            TokenValue::StringLiteral("test"),
            TokenValue::Backtick(),
            TokenValue::Question(),
            TokenValue::Keyword(VAR),
            TokenValue::Identifier("exit_code"),
            TokenValue::Semicolon(),
        ];

        let expected_tree = CodeBlock {
            functions: vec![],
            statements: vec![Statement::Expression(Expression::Execute(
                Box::new(Expression::Command(vec!["test".into()])),
                Some(expressions::CaptureExitCode::Declaration(
                    "exit_code".into(),
                )),
            ))],
        };

        parser_test(tokens, Ok(expected_tree));
    }

    #[test]
    fn should_parse_booleans() {
        let tokens = vec![
            TokenValue::Keyword(VAR),
            TokenValue::Identifier("my_variable"),
            TokenValue::Equals(),
            TokenValue::Keyword(TRUE),
            TokenValue::Semicolon(),
        ];

        let expected_tree = CodeBlock {
            functions: vec![],
            statements: vec![Statement::Declaration(
                Assignment::Simple("my_variable".into()),
                Expression::BooleanLiteral(true),
            )],
        };

        parser_test(tokens, Ok(expected_tree));
    }

    fn parser_test(tokens: Vec<TokenValue>, expected_tree: Result<CodeBlock, ParserError>) {
        let tree = parse(
            tokens
                .into_iter()
                .map(|value| Token::new(value, FilePosition::new(0, 0), FilePosition::new(0, 0))),
        );

        assert_eq!(
            tree, expected_tree,
            "Checking expected tree matches actual tree, actual on left"
        )
    }
}
