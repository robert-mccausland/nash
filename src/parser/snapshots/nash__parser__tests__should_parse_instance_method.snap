---
source: src/parser/mod.rs
expression: parse(tokens).unwrap()
---
statements:
  - Expression:
      operations: []
      first:
        content:
          VariableExpression:
            name:
              value: array
            arguments: ~
        accessors:
          - Variable:
              name:
                value: push
              arguments:
                - operations: []
                  first:
                    content:
                      VariableExpression:
                        name:
                          value: value
                        arguments: ~
                    accessors: []
functions: []
